### PR TITLE
QueryManager: memoize result of getItems, based on items and itemKeys

### PR DESCRIPTION
**Pull request by jsnajdr**
Fixes a `QueryManager.getItems` memoization issue identified in https://github.com/Automattic/wp-calypso/pull/44963#discussion_r472069325

Adds a two-level memoization cache (`items` + `itemKeys`) when computing the `getItems` result.

That should make subsequent `getMedia` selector results `===` to each other more often.

**Todo:**
- [x] add unit tests
- [x] memoize the `PaginatedQueryManager`

### CODE REVIEW

File: client/lib/query-manager/test/index.js [hunk 1]
```diff
@@ -118,6 +118,21 @@ describe( 'QueryManager', () => {
...
+			const queryItems1 = manager.getItems( {} );
+			const allItems2 = manager.getItems();
+			const queryItems2 = manager.getItems( {} );
+
+			expect( allItems1.map( ( item ) => item.ID ) ).toEqual( [ 144, 152 ] );
+			expect( allItems1 ).toBe( allItems2 );
+
+			expect( queryItems1.map( ( item ) => item.ID ) ).toEqual( [ 152 ] );
+			expect( queryItems1 ).toBe( queryItems2 );
+		} );
```
[the remark this answers is not in the document]
Comment (jsnajdr) [reply client/lib/query-manager/test/index.js]: Yes! Added in https://github.com/Automattic/wp-calypso/pull/44986/commits/266343fabadc1e2e4556a59ef87abd8403bef6e0

The `PaginatedQueryManager` tests suite has a similar test, too.

File: client/lib/query-manager/index.js [hunk 2]
```diff
@@ -20,6 +20,41 @@ import QueryKey from './key';
  */
 export const DELETE_PATCH_KEY = '__DELETE';
 
+/*
+ * Construct an array of items from an array of keys and a map of key/item pairs.
+ * There is a two-level memoization cache for the `items` and `itemKeys` params.
+ * `itemKeys` can also be `null`, which means a request to return an array of all items.
+ */
+const itemsCache = new WeakMap();
```
[the remark this answers is not in the document]
Comment (jsnajdr) [reply client/lib/query-manager/index.js:28]: Yes. The `items` object comes from the Redux state tree, and once the Redux state is updated with a new version of `items`, the old version is garbage collected. If we put `items` as a key into `WeakMap`, the record will be automatically removed during GC. That way, the `WeakMap` cache keeps only records for `items` that are referenced from somewhere and are still "alive".

This `WeakMap` memoization is useful for inputs that have unique instances, like arrays or objects. Primitive values cannot be used as keys here. These need to be `Map` and forgetting stale cache records is not that easy.

File: client/lib/query-manager/index.js [hunk 3]
```diff
@@ -20,6 +20,41 @@ import QueryKey from './key';
...
+			resultForAllKeys = values( items );
+			cacheForItems.set( ALL_ITEMS_KEY, resultForAllKeys );
+		}
+		return resultForAllKeys;
+	}
+
+	// compute result from `items` and `itemKeys`, cached for unique `itemKeys` instances
+	let resultForItemKeys = cacheForItems.get( itemKeys );
+	if ( ! resultForItemKeys ) {
+		resultForItemKeys = itemKeys.map( ( itemKey ) => items[ itemKey ] );
```
[the remark this answers is not in the document]
Comment (jsnajdr) [reply client/lib/query-manager/index.js:52]: I wanted the `getItemsForKeys` function to be self-contained, depending only on arguments, without referencing `this` or anything else. That's why I inlined `getItem`.

As I understand it, derived classes are not really supposed to override these methods: the intended extensions points are the `matches`, `sort`, `compare` functions and the `QueryKey` and `DefaultQuery` fields. These are specific for item types like "post", "term" or "theme".

The only exception is `PaginatedQueryManager` that overrides `getItems`. But it seems to be that `QueryManager` and `PaginatedQueryManager` should really be merged into one base class. Most WordPress endpoints are paginated anyway.

**Comment by jsnajdr** [comment]
> Without memoizing the Paginated Query Manager I wonder how often we're actually getting the benefit from this?

Today, I added `PaginatedQueryManager` memoization, too 🙂 

The main motivation for this patch was optimizing the `getMedia` selector results in #44963, which uses the non-paginated method. So, even without support for paginated queries, there would be a benefit.

File: client/lib/query-manager/index.js [hunk 4]
```diff
@@ -20,6 +20,41 @@ import QueryKey from './key';
...
+function getItemsForKeys( items, itemKeys ) {
+	// Get the cache record for the `items` instance, construct a new one if doesn't exist yet.
+	let cacheForItems = itemsCache.get( items );
+	if ( ! cacheForItems ) {
+		cacheForItems = new WeakMap();
+		itemsCache.set( items, cacheForItems );
+	}
+
+	// `itemKeys === null` means a request for array of all items. Cache them with a special key.
+	if ( ! itemKeys ) {
```
[the remark this answers is not in the document]
Comment (jsnajdr) [reply client/lib/query-manager/index.js]: Fixed in https://github.com/Automattic/wp-calypso/pull/44986/commits/a2e78f18efcf5c48d424c1f3ae2c30521396242f

File: client/lib/query-manager/index.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (saramarcondes) [reply client/lib/query-manager/index.js:52]: Gotcha, thanks for the clarification. This makes sense 🙂

**Review by flootr — APPROVED** [review]
works well, thanks 🚢

**Comment by jsnajdr** [comment]
There are e2e failures, and they seem real. When trying to reproduce one of the failures (activity log after publishing from Gutenberg) on calypso.live, the expected result is not there: the just-published post is not present in the activity log.

I'll try to figure out whether the failure is caused by this PR, or is a part of some general e2e deterioration.

**Comment by jsnajdr** [comment]
About the activity log e2e failure:
- the activity log doesn't refresh when visiting it for a second time, i.e., navigating to Activity Log, then to Editor, then back to Activity Log loads the activity log from REST API only once, on the first visit. That's because the Activity Log uses `requestHttpData` to fetch the data, and that API doesn't have a notion of "refresh on mount". The e2e tests rely on the fact that the Activity Log is visited for the first time after publishing a post
- the `ActivityQueryManager` in `lib/query-manager/activity` is not actually used by any code. There probably used to be a Redux + `<QueryActivity>` API for REST fetching, but that's been replaced by a `requestHttpData` getter.
- the test failure happens when the e2e driver tries to click on the sidebar item: the expected navigation never happens:

Screenshot 2020-08-24 at 08 54 50

Looking at other intermittent e2e failures, I see that this is a common failure mode: an innocuous click in sidebar or elsewhere in the UI doesn't get any reaction. It looks like the browser is frozen or something like that.

**Pull request merged**